arch/xtensa/core/xtensa-asm2.c

```diff
@@ -0,0 +1,236 @@
+
+	__asm__ volatile("rsr.interrupt %0" : "=r"(irqs));
+	__asm__ volatile("rsr.intenable %0" : "=r"(ie));
+	printk(" ** Spurious INTERRUPT(s) %p, INTENABLE = %p\n",
+	       (void *)irqs, (void *)ie);
+	_NanoFatalErrorHandler(_NANO_ERR_RESERVED_IRQ, &_default_esf);
+}
+#endif
+
+static void dump_stack(int *stack)
+{
+	int *bsa = *(int **)stack;
+
+	printk(" **  A0 %p  SP %p  A2 %p  A3 %p\n",
+	       (void *)bsa[BSA_A0_OFF/4], ((char *)bsa) + BASE_SAVE_AREA_SIZE,
+	       (void *)bsa[BSA_A2_OFF/4], (void *)bsa[BSA_A3_OFF/4]);
+
+	if (bsa - stack > 4) {
+		printk(" **  A4 %p  A5 %p  A6 %p  A7 %p\n",
+		       (void *)bsa[-4], (void *)bsa[-3],
+		       (void *)bsa[-2], (void *)bsa[-1]);
+	}
+
+	if (bsa - stack > 8) {
+		printk(" **  A8 %p  A9 %p A10 %p A11 %p\n",
+		       (void *)bsa[-8], (void *)bsa[-7],
+		       (void *)bsa[-6], (void *)bsa[-5]);
+	}
+
+	if (bsa - stack > 12) {
+		printk(" ** A12 %p A13 %p A14 %p A15 %p\n",
+		       (void *)bsa[-12], (void *)bsa[-11],
+		       (void *)bsa[-10], (void *)bsa[-9]);
+	}
+
+#if XCHAL_HAVE_LOOPS
+	printk(" ** LBEG %p LEND %p LCOUNT %p\n",
+	       (void *)bsa[BSA_LBEG_OFF/4],
+	       (void *)bsa[BSA_LEND_OFF/4],
+	       (void *)bsa[BSA_LCOUNT_OFF/4]);
+
+#endif
+
+	printk(" ** SAR %p\n", (void *)bsa[BSA_SAR_OFF/4]);
+}
+
+#if CONFIG_XTENSA_ASM2
+static inline void *restore_stack(void *interrupted_stack)
+{
+	if (!_is_preempt(_current)) {
+		return interrupted_stack;
+	}
+
+	int key = irq_lock();
+
+	_current->switch_handle = interrupted_stack;
+	_current = _get_next_ready_thread();
+
+	void *ret = _current->switch_handle;
+	irq_unlock(key);
+
+	_check_stack_sentinel();
+
+	return ret;
+}
+#endif
+
+/* The wrapper code lives here instead of in the python script that
+ * generates _xtensa_handle_one_int*().  Seems cleaner, still kind of
+ * ugly.
+ */
+#define DEF_INT_C_HANDLER(l)				\
+void *xtensa_int##l##_c(void *interrupted_stack)	\
+{							   \
+	int irqs, m;					   \
+	__asm__ volatile("rsr.interrupt %0" : "=r"(irqs)); \
+							   \
+	while ((m = _xtensa_handle_one_int##l(irqs))) {		\
+		irqs ^= m;					\
+		__asm__ volatile("wsr.intclear %0" : : "r"(m)); \
+	}							\
+	return restore_stack(interrupted_stack);		\
+}
+
+DEF_INT_C_HANDLER(2)
+DEF_INT_C_HANDLER(3)
+DEF_INT_C_HANDLER(4)
+DEF_INT_C_HANDLER(5)
+DEF_INT_C_HANDLER(6)
+DEF_INT_C_HANDLER(7)
+
+static inline DEF_INT_C_HANDLER(1)
+
+/* C handler for level 1 exceptions/interrupts.  Hooked from the
+ * DEF_EXCINT 1 vector declaration in assembly code.  This one looks
+ * different because exceptions and interrupts land at the same
+ * vector; other interrupt levels have their own vectors.
+ */
+void *xtensa_excint1_c(int *interrupted_stack)
+{
+	int cause, vaddr, *bsa = *(int **)interrupted_stack;
+
+	__asm__ volatile("rsr.exccause %0" : "=r"(cause));
+
+	if (cause == EXCCAUSE_LEVEL1_INTERRUPT) {
+
+		return xtensa_int1_c(interrupted_stack);
+
+	} else if (cause == EXCCAUSE_SYSCALL) {
+
+		/* Just report it to the console for now */
```

Comment:
ESP32 has a PID controller.  I initially thought it was related to PID in the sense of control theory, but it's in fact similar in spirit to protection rings on x86.  I don't know how it relates to the generic Xtensa syscall interrupts.